### PR TITLE
Added functionality to output to parquet 

### PR DESCRIPTION
**Pull request by karnesh**
PR to add functionality to output flow, velocity and depth to parquet file. The output parquet is a timeseries data(details in Notes). It is required for the [TEEHR](https://github.com/RTIInternational/teehr) input. TEEHR comprises of set of tools for hydrologic model and forecast evaluation. Storing the t-route output in parquet format will lead to efficient query of the data. Also, it will help in automation by connecting TEEHR with NextGen water model. [CIROH](https://ciroh.ua.edu) along with Lynker has developed a containerized version of NextGen National Water Model [NextGen In A Box (NGIAB)](https://github.com/CIROH-UA/NGIAB-CloudInfra). TEEHR uses DuckDB to query the parquet output from NGIAB stored on cloud.  

## Additions

- Added functionality to output write flow, velocity and depth to parquet format.

## Changes

- Changed yaml input file to include parquet output format
```    
parquet_output:
    #---------
    parquet_output_folder: output/
    configuration: short_range
    prefix_ids: nex
```

## Testing

1. The modified code is tested by executing LowerColorado_TX demonstration test. 
LowerColorado_TX

## Screenshots
- Here is a screenshot of successful compilation.
compile

- Parquet output timeserie data from LowerColorado_TX demonstration test.
parquet

## Notes

- The `flowveldepth` dateframe is modified to create a timeseries containing following variables: 
```
location_id: string
value: double
value_time: timestamp[us]
variable_name: string
configuration: string
units: string
reference_time: timestamp[us]
```
- Location_id variable contains nexus IDs and has 'nex-' prepended to it. 
- Configuration (short range, medium range, long range etc.) has to be entered by the user in the input yaml file. 

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### CODE REVIEW

File: src/troute-nwm/src/nwm_routing/output.py [hunk 1]
```diff
@@ -29,42 +29,84 @@ def _reindex_lake_to_link_id(target_df, crosswalk):
...
+    timeseries_df['value_time'] = (start_datetime + pd.to_timedelta(timeseries_df['timestep'] * dt, unit='s'))
+    variable_to_name_map = {"q": "streamflow", "d": "depth", "v": "velocity"}
+    timeseries_df["variable_name"] = timeseries_df["variable"].map(variable_to_name_map)
+    timeseries_df.drop(['var', 'timestep', 'variable'], axis=1, inplace=True)
+    timeseries_df['configuration'] = configuration
+    variable_to_units_map = {"streamflow": "m3/s", "velocity": "m/s", "depth": "m"}
+    timeseries_df['units'] = timeseries_df['variable_name'].map(variable_to_units_map)
+    timeseries_df['reference_time'] = start_datetime.date()
+    timeseries_df['location_id'] = timeseries_df['location_id'].astype('string')
+    timeseries_df['location_id'] = 'nex-' + timeseries_df['location_id']
```
Comment (hellkite500) [inline src/troute-nwm/src/nwm_routing/output.py:84]: I don't think this is generically appropriate to label all locations in the routing data frame with a `nex-` prefix.  This *may* be applicable to routing which uses a `hy_features network`, but wouldn't be applicable to one which uses a `nhd network`.  Is there any guarantee before this point that this function only gets applied to `hy_features` network results?  Also, even within a `hy_features` network, the routed segments this ouptut relates to is for the `waterbodies` which typically have a `wb-` prefix in the hydrofabric identifiers, not `nex`.  The `nexus` and `waterbody` features are related, but destinct concepts.

File: src/troute-nwm/src/nwm_routing/output.py [hunk 2]
```diff
@@ -29,42 +29,84 @@ def _reindex_lake_to_link_id(target_df, crosswalk):
...
+    - configuration: configuration (for instance- short_range, medium_range)
+
+    Returns:
+    --------
+    - timeseries_df (DataFrame): Converted timeseries data frame
+    '''
+
+    df.index.name = 'location_id'
+    df.reset_index(inplace=True)
+    timeseries_df = df.melt(id_vars=['location_id'], var_name='var')
```
Comment (hellkite500) [inline src/troute-nwm/src/nwm_routing/output.py:70]: I think if you use `value_vars=[ 'q', 'v', 'd' ]` as a kwarg to melt, you might have an easier time extracting from un-pivoted table?

The below method seems like there should be a better way besides casting to string, manipulating the string, and recasting to numeric/datetime types. 

I'm not sure exactly what the `df` looks like that is trying to be manipulated at this point, but I would try to consider a different method(s) for extracting the needed data from it.

If for some reason this is the only way, then please comment this implementation to describe what the state of the df is and why this is the way it needs to be manipulated.

File: src/troute-nwm/src/nwm_routing/output.py [hunk 3]
```diff
@@ -335,12 +383,54 @@ def nwm_output_generator(
...
+
+        # no parquet_output_segments are specified, then write results for all segments
+        if not parquet_output_segments:
+            parquet_output_segments = flowveldepth.index
+
+        flowveldepth = flowveldepth.sort_index()
+        timeseries_df = _parquet_output_format_converter(flowveldepth, restart_parameters.get("start_datetime"), dt,
+                                                         output_parameters["parquet_output"].get("configuration"))
+
+        parquet_output_segments_str = ['nex-' + str(segment) for segment in parquet_output_segments]
```
Comment (hellkite500) [inline src/troute-nwm/src/nwm_routing/output.py:416]: another use of `nex-` prefix that may not be generically appropriate.

**Review by hellkite500 — CHANGES_REQUESTED** [review]
In general, I like the idea of adding this as an option for possible output formats.  There are some details of the implementation that need to be discussed/considered a bit I think, however.

Also, a more general note, but it configuring your local dev environment git to ignore whitespace changes may be helpful on PR's like this which touch many files.  It looks like some auto formatting was applied, and the formatting changes can make review more challenging when trying to see all the places in the code which have changed in functionality.  If you want to contribute formatting changes, I would suggest putting them all into a single commit on a PR, or an independent PR.

File: src/troute-nwm/src/nwm_routing/output.py (hunk 3 above)
Comment (jameshalgren) [reply src/troute-nwm/src/nwm_routing/output.py:416]: Would suggest we make it a variable with a default argument in the wrapping function and a value in the yaml. 
`parquet_output_segments_str = [prefix_str + str(segment) for segment in parquet_output_segments]`

We would need to do something more comprehensive to update T-Route to comprehend the naming/labeling of the IDs...

File: src/troute-nwm/src/nwm_routing/output.py (hunk 2 above)
Comment (karnesh) [reply src/troute-nwm/src/nwm_routing/output.py:70]: I looked into options and this seems to be the only way. I had to manipulate the strings and cannot use `value_vars=[ 'q', 'v', 'd' ]` as a kwarg to melt because of the `df` format. Please look at the screenshot below.

parquet t-route

Each column name consists of a time step and a variable name (`q`, `v` or `d`) in string format. The time steps values are used to get the value_time.

File: src/troute-nwm/src/nwm_routing/output.py (hunk 3 above)
Comment (karnesh) [reply src/troute-nwm/src/nwm_routing/output.py]: I have modified the PR to add user defined value in yaml file for the prefix string.

File: src/troute-nwm/src/nwm_routing/output.py [hunk 4]
```diff
@@ -29,25 +29,25 @@ def _reindex_lake_to_link_id(target_df, crosswalk):
...
 
     # (re) set the target_df index
-    target_df.set_index(idxs, inplace=True)
+    target_df.set_index(idxs, inplace = True)
 
     return target_df
 
 
-def _parquet_output_format_converter(df, start_datetime, dt, configuration):
+def _parquet_output_format_converter(df, start_datetime, dt, configuration, prefix_ids):
```
Comment (jameshalgren) [inline src/troute-nwm/src/nwm_routing/output.py:50]: We can drop most of these format only changes to keep the PR super clean.

Comment (karnesh) [reply src/troute-nwm/src/nwm_routing/output.py]: I have reverted the format changes.

**Comment by shorvath-noaa** [comment]
@hellkite500 @jameshalgren does this PR still need changes in your opinions or can we review/approve?

**Comment by jameshalgren** [comment]
> @hellkite500 @jameshalgren does this PR still need changes in your opinions or can we review/approve?

Please proceed with final review.

File: src/troute-config/troute/config/output_parameters.py [hunk 5]
```diff
@@ -4,13 +4,16 @@
 from typing_extensions import Literal
 from .types import FilePath, DirectoryPath
 
-streamOutput_allowedTypes = Literal['.csv', '.nc', '.pkl']
+streamOutput_allowedTypes = Literal['.csv', '.nc', '.pkl', '.parquet']
```
Comment (shorvath-noaa) [inline src/troute-config/troute/config/output_parameters.py:7]: `streamOutput_allowedTypes` is specific for the `stream_output` class of parameters. It looks like you've created a new class, `ParquetOutput` that doesn't use this. Is this change needed?

**Review by shorvath-noaa — CHANGES_REQUESTED** [review]
Apologies for the slow review. Can you rebase this on the master branch to resolve conflicts?

File: src/troute-config/troute/config/output_parameters.py (hunk 5 above)
Comment (karnesh) [reply src/troute-config/troute/config/output_parameters.py]: Thank you for pointing this out. I have removed `.parquet` from `streamOutput_allowedTypes`

**Comment by karnesh** [comment]
I have rebased the PR to resolve conflicts.

**Comment by shorvath-noaa** [comment]
@karnesh it looks like you are missing [this line](https://github.com/NOAA-OWP/t-route/blob/e6a1b85b6b4d1172ac6a75939001a9d486f3dedb/src/troute-nwm/src/nwm_routing/output.py#L62) in your PR. I think this is why it is failing the GitActions tests. Can you add it in?

**Review by shorvath-noaa — APPROVED** [review]
